include/boost/url/detail/segments_encoded_iterator_base.hpp

```diff
@@ -0,0 +1,92 @@
+//
+// Copyright (c) 2019 Vinnie Falco (vinnie.falco@gmail.com)
+// Copyright (c) 2022 Alan de Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_DETAIL_SEGMENTS_ENCODED_ITERATOR_BASE_HPP
+#define BOOST_URL_DETAIL_SEGMENTS_ENCODED_ITERATOR_BASE_HPP
+
+#include <string>
+#include <boost/url/string_view.hpp>
+
+namespace boost {
+namespace urls {
+namespace detail {
+
+struct segments_encoded_iterator_base
+{
+    std::size_t i_ = 0;
+    string_view s_;
+    char const* begin_ = nullptr;
+    char const* pos_ = nullptr;
+    char const* next_ = nullptr;
+    char const* end_ = nullptr;
+
+    segments_encoded_iterator_base(
+        string_view s,
+        std::size_t nseg) noexcept;
+
+    // end ctor
+    segments_encoded_iterator_base(
+        string_view s,
+        std::size_t nseg,
+        int) noexcept;
+
+    using value_type = std::string;
+    using reference = string_view;
+    using pointer = void const*;
+    using difference_type = std::ptrdiff_t;
+    using iterator_category =
+        std::bidirectional_iterator_tag;
+
+    segments_encoded_iterator_base() = default;
+
+    segments_encoded_iterator_base(
+        segments_encoded_iterator_base const&) noexcept = default;
+
+    segments_encoded_iterator_base& operator=(
+        segments_encoded_iterator_base const&) noexcept = default;
+
+    reference
+    operator*() const noexcept
+    {
+        return s_;
```

Comment:
there's no point to this function, you should just `return s_` in the enclosing class